### PR TITLE
added support for legacy zeppelin notebooks

### PR DESCRIPTION
**Pull request by comloo**

### CODE REVIEW

File: zeppelin/cli.py [hunk 1]
```diff
@@ -19,9 +23,23 @@ def main():
...
 
-    zeppelin_converter = ZeppelinConverter(args.in_filename, args.out_filename,
-                                           directory, 'anonymous', 'N/A', 'N/A')
-    zeppelin_converter.convert()
+    try:
+        with open(args.in_filename, 'rb') as raw:
+            t = json.load(raw)
+            full_path = os.path.join(directory, args.out_filename + '.md')
+            with open(full_path, 'w') as fout:
+                if 'results' in t['paragraphs'][0]:
```
Comment (maurodoglio) [inline zeppelin/cli.py]: Can you put this logic in a function please? Even if it's just a one line, I think it's worth it for example for testing

File: zeppelin/cli.py [hunk 2]
```diff
@@ -19,9 +23,23 @@ def main():
...
+                    zeppelin_converter = nc(args.in_filename, args.out_filename,
+                                            directory, 'anonymous', 'N/A', 'N/A')
```
Comment (maurodoglio) [inline zeppelin/cli.py]: 'anonymous', 'N/A', etc. should be the default values. Please use keyword arguments in the function definition instead.

File: zeppelin/cli.py [hunk 3]
```diff
@@ -19,9 +23,23 @@ def main():
...
+                else:
+                    zeppelin_converter = lc(args.in_filename, args.out_filename,
+                                            directory, 'anonymous', 'N/A', 'N/A')
+
+                zeppelin_converter.convert(t, fout)
+
+    except ValueError as err:
+        print(err)
```
Comment (maurodoglio) [inline zeppelin/cli.py]: Catching and printing the exception doesn't add much value compared to letting the exception raise. Could you instead print a human-readable message ?

File: zeppelin/converter.py [hunk 4]
```diff
@@ -38,6 +35,7 @@ def build_header(self, title):
         self.out = header + self.out
 
     def build_markdown(self, lang, body):
+        """Append output with text body."""
```
Comment (maurodoglio) [inline zeppelin/converter.py]: Could you please reword this?

File: zeppelin/legacy_converter.py [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+        """
+        result = re.search('xml version', msg)
+        if result is None:
+            return
+
+        self.index += 1
+        images_path = 'images'
+
+        if self.directory:
+            images_path = self.directory + '/' + images_path
```
Comment (maurodoglio) [inline zeppelin/legacy_converter.py]: please use os.path.join instead

File: zeppelin/legacy_converter.py [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+
+    def build_text(self, msg):
+        """Add text to output array."""
+        self.out.append(msg)
+
+    def build_table(self, msg):
+        """Format each row of the table."""
+        rows = msg.split('\n')
+        if rows:
+            header_row = rows[0]
```
Comment (maurodoglio) [inline zeppelin/legacy_converter.py]: now that we use python3 we can make unpack lists using the star operator:
```python
header_row, *body_rows = rows
```

File: zeppelin/legacy_converter.py [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+            body_rows = rows[1:]
+            self.create_md_row(header_row, True)
+            for row in body_rows:
+                self.create_md_row(row)
+
+    def process_results(self, paragraph):
+        """Output options.
```
Comment (maurodoglio) [inline zeppelin/legacy_converter.py]: I would reword this to:
```python
"""Routes Zeppelin output types to corresponding handlers

To add support for other output types, add the file type to the dictionary
and create the necessary function to handle it.
"""
```

File: zeppelin/new_converter.py [hunk 8]
```diff
@@ -0,0 +1,98 @@
+import re
+import os
+import base64
+from .converter import ZeppelinConverter
+
+
+class NewConverter(ZeppelinConverter):
+    """NewConverter class.
```
Comment (maurodoglio) [inline zeppelin/new_converter.py]: Please remove the first line as it's meaningless. You can you the rest for the short description.

File: zeppelin/new_converter.py [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+        rows = msg['data'].split('\n')
+        if rows:
+            header_row = rows[0]
+            body_rows = rows[1:]
+            self.create_md_row(header_row, True)
+            for row in body_rows:
+                self.create_md_row(row)
+
+    def process_results(self, paragraph):
+        """Output options.
```
Comment (maurodoglio) [inline zeppelin/new_converter.py]: Here as well

File: setup.cfg [hunk 10]
```diff
@@ -3,3 +3,4 @@ test=pytest
 
 [flake8]
 max-line-length=100
+ignore=D102,D100,D103,D104,E211
```
Comment (maurodoglio) [inline setup.cfg]: I got the same errors locally. Did you install flake8-docstrings? Without that installed this class of errors won't show up.

File: README.md [hunk 11]
```diff
@@ -1,5 +1,9 @@
 # Zeppelin Converter
 
+### Requirements
+
+- Python 3.\*
```
Comment (maurodoglio) [inline README.md]: You should mention that it's tested on python 3.6, but it may work on other versions.

File: README.md [hunk 12]
```diff
@@ -1,5 +1,9 @@
...
+- Cairo: `brew install cairo`
```
Comment (maurodoglio) [inline README.md]: This works only on OSX. Instead of giving the exact command, you could explain what Cairo is and why it's needed and suggest to install it using watchever package system is available on the user machine.

File: zeppelin/cli.py [hunk 13]
```diff
@@ -2,9 +2,16 @@
...
 import json
 import sys
-from .new_converter import NewConverter as nc
-from .legacy_converter import LegacyConverter as lc
+from .new_converter import NewConverter
+from .legacy_converter import LegacyConverter
 
+def check_version(text, args, directory):
+    if 'results' in text['paragraphs'][0]:
+        return NewConverter(args.in_filename, args.out_filename,
```
Comment (maurodoglio) [inline zeppelin/cli.py]: This function should only check the zeppelin version, like its name says. You could even rename the function `is_new_format` since that's exactly what it does. Please leave the rest of the code in the calling function.

File: zeppelin/new_converter.py [hunk 14]
```diff
@@ -0,0 +1,92 @@
+import re
+import os
+import base64
+from .converter import ZeppelinConverter
+
+
+class NewConverter(ZeppelinConverter):
+    """NewConverter converts Zeppelin version 0.7.1 notebooks to Markdown."""
+
+    MD_EXT = '.md'
```
Comment (maurodoglio) [inline zeppelin/new_converter.py]: This should probably go in the base class. Actually, what about renaming the base class MarkdownConverter? The `zeppelin` part can be omitted since that's already the module name.

File: zeppelin/new_converter.py [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+        """Generate the body for the Markdown file.
+
+        - processes each json block one by one
+        - for each block
+            - process the input by detecting the editor language
+            - print the input
+            - process the output by detecting the output format
+            - print the output
+        """
+        for paragraph in text['paragraphs']:
```
Comment (maurodoglio) [inline zeppelin/new_converter.py]: I would be tempted to convert this IFs into a dict(paragraph_key -> handler) and iterate on that. What do you think?

File: zeppelin/new_converter.py [hunk 16]
```diff
@@ -0,0 +1,92 @@
...
+            mode = paragraph['config']['editorMode'].split('/')[-1]
+            if paragraph['results']['msg']:
+                msg = paragraph['results']['msg'][0]
+                if mode not in ('text', 'markdown'):
+                    output_options[msg['type']](msg)
+
+    def build_markdown_body(self, text):
+        """Generate the body for the Markdown file.
+
+        - processes each json block one by one
```
Comment (maurodoglio) [inline zeppelin/new_converter.py]: I don't think this part of the docstrings applies anymore

File: zeppelin/new_converter.py (hunk 15 above)
Comment (maurodoglio) [inline zeppelin/new_converter.py]: This function seems to do the same as its counterpart in LegacyConverter, could you put it in the parent class?

File: zeppelin/legacy_converter.py [hunk 17]
```diff
@@ -0,0 +1,88 @@
...
+        if not os.path.isdir(images_path):
+            os.makedirs(images_path)
+
+        with open('{0}/output_{1}.png'.format(images_path, self.index), 'wb') as fh:
+            cairosvg.svg2png(bytestring=msg.encode('utf-8'), write_to=fh)
+
+        self.out.append(
+            '\n![png]({0}/output_{1}.png)\n'.format(images_path, self.index))
+
+    def build_text(self, msg):
```
Comment (maurodoglio) [inline zeppelin/legacy_converter.py]: This functio ncan be factored out into the base class

File: zeppelin/new_converter.py (hunk 15 above)
Comment (comloo) [inline zeppelin/new_converter.py]: @Second comment: In LegacyConverter, its if 'result' in paragraph, whereas in NewConverter, its if 'results' in paragraph. (It's actually one of the few ways to distinguish between versions ): )

File: zeppelin/legacy_converter.py (hunk 17 above)
Comment (comloo) [inline zeppelin/legacy_converter.py]: In LegacyConverter, it's `self.out.append(msg)`, whereas in NewConverter, its `self.out.append(msg['data'])`. A bit different; not sure exactly how to proceed with this, since the JSON formats are a bit different.

File: zeppelin/new_converter.py (hunk 15 above)
Comment (maurodoglio) [inline zeppelin/new_converter.py]: ok so what about having a class attribute "result_key" to the base class and only override this attribute in one of the subclasses ?

File: zeppelin/legacy_converter.py (hunk 17 above)
Comment (maurodoglio) [inline zeppelin/legacy_converter.py]: If I understand correctly, in process_results you could pass msg['data'] to the output_options handlers instead of msg.

Comment (maurodoglio) [inline zeppelin/legacy_converter.py]: If you do that, you can also have a unique implementation of build_table in the base class

File: setup.cfg (hunk 10 above)
Comment (comloo) [inline setup.cfg]: forgot to do this! thanks for the heads up :)
